llvm/lib/IR/RepoTicket.cpp

```diff
@@ -116,78 +125,135 @@ static GODigestState accumulateGOFinalDigestOrInitialDigestAndGetGODigestState(
   return accumulateGOInitialDigestAndGetGODigestState(GO, GOHash, GOIMap);
 }
 
-// Update the GO's hash value by adding the hash of its dependents.
-template <typename Function>
-static bool updateDigestUseDependenciesAndContributions(
-    const GlobalObject *GO, MD5 &GOHash, GOStateMap &Visited, GOInfoMap &GOIMap,
-    Function AccumulateGODigest) {
+// Update the digest of an invidual GO incorporating the hashes of all its
+// dependencies and contributions.
+// \param GO The global object whose digest is to be computed.
+// \param Visited A map of GO to a unique number in the dependencies or
+//        contributions graph.
+// \param GOIMap A map of GO to GOInfo
+// \param GOHashCache Used to record memoized hashes.
+// \param AccumulateGODigest A function is used to accumulate the GO's digest.
+// \returns a tuple containing the state information which includes 1) return
+// true if the module M has been changed, 2) a numerical identifier which will
+// be used if GO loops back to itself in future and 3) the GO's digest.
+template <typename FunctionType>
+static auto updateDigestUseDependenciesAndContributions(
```

Comment:
I think that this function has become too complicated and should probably be broken down. It is 84 lines long. Could the cache be implemented as a class which could then simplify:

```
  // If the hash has been memoized, we can return the result immediately.
  const auto TablePos = GOHashCache.find(GO);
  if (TablePos != GOHashCache.end()) {
    ++VisitedCachedGOTimes;
    LLVM_DEBUG(dbgs() << "Returning pre-computed hash for \"" << GO->getName()
                      << "\"\n");
    return std::make_tuple(true, NumVisited, TablePos->second);
  }
```
to something like
```
  // If the hash has been memoized, return the result immediately.
  std::tie(Hit, Hash) = GOHashCache.get(GO)
  if (Hit) {
    return std::make_tuple(true, NumVisited, Hash);
  }
```
I'm not sure whether the LLVM_DEBUG is more useful inside or outside the cache.

Reply:
I have put the hash calculation in a Class named `HashCache`, Please see the latest commit for the details.